modules/giapi/src/main/scala/giapi/client/CommandSenderClient.scala

```diff
@@ -0,0 +1,199 @@
+                // We need to close the consumer after we get a response or we will leak
+                // resources and get meessages for every command
+                def safeClose(): Unit =
+                  try
+                    responseConsumer.close()
+                  catch {
+                    case _: Exception =>
+                  }
+
+                // Listen to a JMS message, parse it and return the result to the caller
+                responseConsumer.setMessageListener(new MessageListener {
+                  override def onMessage(response: Message): Unit =
+                    if (response.getJMSCorrelationID === id) {
+                      parseResponse(response) match {
+                        case Some(r @ CommandResult(Response.ACCEPTED))  =>
+                          safeClose()
+                          cb(Right(r))
+                        case Some(CommandResult(Response.STARTED))       =>
+                        // We have to wait, do nothing
+                        case Some(r @ CommandResult(Response.COMPLETED)) =>
+                          safeClose()
+                          cb(Right(r))
+                        // Cannot happen, but required by the compiler
+                        case Some(CommandResult(Response.ERROR))         =>
+                          safeClose()
+                        case Some(CommandResult(Response.NOANSWER))      =>
+                          safeClose()
+                          cb(
+                            Left(
+                              CommandResultException("No answer from the instrument")
+                            )
+                          )
+                        case Some(CommandResultException(_, msg))        =>
+                          safeClose()
+                          cb(Left(new GiapiException(msg)))
+                        case None                                        =>
+                          safeClose()
+                          cb(
+                            Left(
+                              new GiapiException("Incorrect response, possible coding error")
+                            )
+                          )
+                      }
+                    }
+                })
+                responseConsumer
+              }
+          _     <- Sync[F]
+                     .delay {
+                       // Send the message to the GMP
+                       val messageBuilder =
+                         MessageBuilderFactory.newMessageBuilder(command.toGiapi, id)
+                       val msg            = session.createMapMessage()
+                       messageBuilder.constructMessageBody(msg)
+                       msg.setJMSReplyTo(dest)
+                       producer.send(msg)
+                     }
+                     .handleError { e =>
+                       cb(Left(e))
+                     }
+        } yield Some(Sync[F].delay(reply.close()).attempt.void)
```

Comment:
```suggestion
          responseConsumer <- Sync[F].delay(session.createConsumer(dest))
          _ <-
            Sync[F]
              .delay {
                // Listen to a JMS message, parse it and return the result to the caller
                responseConsumer.setMessageListener(new MessageListener {
                  override def onMessage(response: Message): Unit =
                    if (response.getJMSCorrelationID === id) {
                      parseResponse(response) match {
                        case Some(r @ CommandResult(Response.ACCEPTED))  =>
                          cb(Right(r))
                        case Some(CommandResult(Response.STARTED))       =>
                        // We have to wait, do nothing
                        case Some(r @ CommandResult(Response.COMPLETED)) =>
                          cb(Right(r))
                        // Cannot happen, but required by the compiler
                        case Some(CommandResult(Response.ERROR))         =>
                          ()
                        case Some(CommandResult(Response.NOANSWER))      =>
                          cb(
                            Left(
                              CommandResultException("No answer from the instrument")
                            )
                          )
                        case Some(CommandResultException(_, msg))        =>
                          cb(Left(new GiapiException(msg)))
                        case None                                        =>
                          cb(
                            Left(
                              new GiapiException("Incorrect response, possible coding error")
                            )
                          )
                      }
                    }
                })
              }
          _     <- Sync[F]
                     .delay {
                       // Send the message to the GMP
                       val messageBuilder =
                         MessageBuilderFactory.newMessageBuilder(command.toGiapi, id)
                       val msg            = session.createMapMessage()
                       messageBuilder.constructMessageBody(msg)
                       msg.setJMSReplyTo(dest)
                       producer.send(msg)
                     }
                     .handleError { e =>
                       cb(Left(e))
                     }
        } yield Some(Sync[F].delay(responseConsumer.close()).attempt.void)
```

Reply:
I think `close()` will be called right after each `cb` invocation, wasn't it being called twice?

You might want to throw though in case of `Response.ERROR` to ensure that the effect completes and cleans up.

Reply:
Sorry, please ignore the above suggestion. I thought the returned cleanup callback was called always, but it's only called on fiber cancellation.